src/test/java/org/apache/commons/compress/utils/SeekableInMemoryByteChannelTest.java

```diff
@@ -155,15 +155,97 @@ public void shouldTruncateContentsProperly() throws IOException {
+                // - Position doesn?t change
+                { 9L, 9L, 9L },
+
+                // NewSize < Position == Size :
+                // - Size changes to NewSize
+                // - Position changes to NewSize
+                { 4L, 9L, 4L },
+
+                // Position == Size < NewSize :
+                // - Size changes to NewSize and you get undefined data between
+                // - Position doesn?t change
+                { 9L, 12L, 9L },
+
+                // NewSize == Position < Size :
+                // - Size changes to NewSize
+                // - Position doesn?t change
+                { 4L, 4L, 4L },
+
+                // Size < Position == NewSize :
+                // - Size changes to NewSize and you get undefined data between
+                // - Position doesn?t change
+                { 12L, 12L, 12L },
+
+                // Position < Size == NewSize :
+                // - Size doesn?t change
+                // - Position doesn?t change
+                { 4L, 9L, 4L },
+
+                // NewSize == Size < Position :
+                // - Size doesn?t change
+                // - Position changes to NewSize
+                { 12L, 9L, 9L },
+
+
+                // NewSize < Position < Size :
+                // - Size changes to NewSize
+                // - Position changes to NewSize
+                { 6L, 4L, 4L },
+
+                // Position < NewSize < Size :
+                // - Size changes to NewSize
+                // - Position doesn?t change
+                { 4L, 6L, 4L },
+
+                // Position < Size < NewSize :
+                // - Size changes to NewSize and you get undefined data between
+                // - Position doesn?t change
+                { 4L, 12L, 4L },
+
+                // NewSize < Size < Position :
+                // - Size changes to NewSize
+                // - Position changes to NewSize
+                { 12L, 4L, 4L },
+
+                // Size < NewSize < Position :
+                // - Size changes to NewSize and you get undefined data between
+                // - Position changes to NewSize
+                { 12L, 10L, 10L },
+
+                // Size < Position < NewSize :
+                // - Size changes to NewSize and you get undefined data between
+                // - Position doesn?t change
+                { 10L, 12L, 10L }
+
+        };
+
+        for (long[] testCase : testCases) {
+            try {
+                long position = testCase[0];
+                long newSize = testCase[1];
+                long expectedPosition = testCase[2];
+
+                //given
+                try (SeekableInMemoryByteChannel c = new SeekableInMemoryByteChannel(testData)) {
+                    //when
+                    c.position(position);
+                    c.truncate(newSize);
+                    //then
+                    assertEquals(expectedPosition, c.position());
+                    assertEquals(newSize, c.size());
```

Comment:
I think this is not right. Basing on the java docs, I think it should be like this :
```
//when
long originalSize = c.size();
c.position(position);
c.truncate(newSize);
//then
assertEquals(expectedPosition, c.position());
long expectedSize = newSize < originalSize ? newSize : originalSize;
assertEquals(expectedSize, c.size());
```

Reply:
You might be right about that. If that's the case I'd prefer to put expectedSize into the array though, otherwise it's dangerously like replicating the logic in the class being tested.

Reply:
> if the position is past the end and the truncation point, then the position is updated to the truncation point, but the size is not updated to it.

I don't see any problems with the `size`. I think that's what the name `truncate` means. It's called `truncate`, not `resize` or some other name, which means the size can only become smaller when this method is called. If the `newSize` is bigger than the `size`, I think there's nothing needed to do on the `size`.
I think the problem is about the `position`, not the `size`.